File: prism-kotlin-sdk/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/identity/DIDSuffix.kt
package io.iohk.atala.prism.kotlin.identity

import io.iohk.atala.prism.kotlin.crypto.SHA256Digest

data class DIDSuffix(val value: String) {
    companion object {
        private val suffixRegex = Regex("[:A-Za-z0-9_-]+$")

        fun fromString(suffix: String): DIDSuffix =
            apply(suffix)

        @ExperimentalUnsignedTypes
        fun fromDigest(digest: SHA256Digest): DIDSuffix =
            apply(digest.hexValue())

        private fun apply(didSuffix: String): DIDSuffix {
            require(suffixRegex.matches(didSuffix)) {
                "Invalid DID Suffix format: $didSuffix"
            }
            return DIDSuffix(didSuffix)
        }
    }

    override fun toString(): String = value
}

File: prism-sdk/identity/shared/src/main/scala/io/iohk/atala/prism/identity/DIDSuffix.scala
package io.iohk.atala.prism.identity

import io.iohk.atala.prism.crypto.Sha256Digest

import scala.util.Try

final class DIDSuffix private (val value: String) extends AnyVal {
  override def toString: String = value
}

object DIDSuffix {
  def fromString(suffix: String): Option[DIDSuffix] = {
    Try(DIDSuffix(suffix)).toOption
  }

  def unsafeFromDigest(digest: Sha256Digest): DIDSuffix = apply(digest.hexValue)

  def unsafeFromString(didSuffix: String): DIDSuffix = apply(didSuffix)

  private def apply(didSuffix: String): DIDSuffix = {
    require(DID.suffixRegex.pattern.matcher(didSuffix).matches(), s"invalid DID Suffix format: $didSuffix")
    new DIDSuffix(didSuffix)
  }
}
